src/sp_cookie_encryption.c

```diff
@@ -88,64 +92,54 @@ PHP_FUNCTION(sp_setcookie) {
 
   /* Shall we encrypt the cookie's value? */
   if (cookie_node && cookie_node->encrypt && value) {
-    zend_string *encrypted_data = encrypt_data(value);
-    ZVAL_STR_COPY(&params[1], encrypted_data);
-    zend_string_release(encrypted_data);
-  } else if (value) {
-    ZVAL_STR_COPY(&params[1], value);
+    value_enc = encrypt_data(value);
   }
 
-  ZVAL_STRING(&func_name, "setcookie");
-  ZVAL_STR_COPY(&params[0], name);
-  ZVAL_LONG(&params[2], expires);
-  if (domain) {
-    ZVAL_STR_COPY(&params[4], domain);
-  }
-  if (secure) {
-    ZVAL_LONG(&params[5], secure);
-  }
-  if (httponly) {
-    ZVAL_LONG(&params[6], httponly);
-  }
 
-  /* param[3](path) is concatenated to path= and is not filtered in PHP < 7.3
-     we can inject the samesite parameter here
-     TODO find another solution with 7.3 */
   if (cookie_node && cookie_node->samesite) {
     if (!path) {
       path = zend_string_init("", 0, 0);
     }
+#if PHP_VERSION_ID < 70300
     cookie_samesite = (cookie_node->samesite == lax)
                           ? SAMESITE_COOKIE_FORMAT SP_TOKEN_SAMESITE_LAX
                           : SAMESITE_COOKIE_FORMAT SP_TOKEN_SAMESITE_STRICT;
+
     /* Concatenating everything, as is in PHP internals */
-    samesite = zend_string_init(ZSTR_VAL(path), ZSTR_LEN(path), 0);
-    samesite = zend_string_extend(
-        samesite, ZSTR_LEN(path) + strlen(cookie_samesite) + 1, 0);
-    memcpy(ZSTR_VAL(samesite) + ZSTR_LEN(path), cookie_samesite,
+    path_samesite = zend_string_init(ZSTR_VAL(path), ZSTR_LEN(path), 0);
+    path_samesite = zend_string_extend(
+        path_samesite, ZSTR_LEN(path) + strlen(cookie_samesite) + 1, 0);
+    memcpy(ZSTR_VAL(path_samesite) + ZSTR_LEN(path), cookie_samesite,
            strlen(cookie_samesite) + 1);
-    ZVAL_STR_COPY(&params[3], samesite);
-  } else if (path) {
-    ZVAL_STR_COPY(&params[3], path);
+#else
+    cookie_samesite = (cookie_node->samesite == lax)
+                          ? SP_TOKEN_SAMESITE_LAX
+                          : SP_TOKEN_SAMESITE_STRICT;
+
+    samesite = zend_string_init(cookie_samesite, strlen(cookie_samesite), 0);
+#endif
+  }
+
+
+#if PHP_VERSION_ID < 70300
+  if (php_setcookie(name, (value_enc ? value_enc : value), expires, (path_samesite ? path_samesite : path), domain, secure, 1, httponly)) {
+#else
+  if (php_setcookie(name, (value_enc ? value_enc : value), expires, path, domain, secure, httponly, samesite, 1)) {
```

Comment:
What if another module wants to hook `setcookie`?

Reply:
indeed.. will be an issue (depending of load order)

Reply:
I'm not sure that we want to prevent other extension hooking the function from working :/